docs(adr): ADR-009: Supply module (#786)
desmos-labs/desmos: docs/architecture/adr-009-supply-module.md

# ADR 009: Supply module

## Changelog

- March 21, 2022: Initial draft.

## Status

DRAFTED

## Abstract

> This ADR defines the `x/supply` module which will expose a set of APIs that will be called by data aggregator websites (such as Coingecko and CoinMarketCap)
> in order to fetch updated data about $DSM supply.

Reviewer: ```suggestion
This ADR defines the `x/supply` module which will expose a set of APIs that can be called by data aggregator websites (such as Coingecko and CoinMarketCap) in order to fetch updated data about a specific token supply.
```

## Context

> Currently, inside our [CoinGecko](https://www.coingecko.com/en/coins/desmos) and [CoinMarketCap](https://coinmarketcap.com/currencies/desmos/) 
> some important information (current and total supply) are missing or not correctly updated. To solve this, we can implement a
> series of APIs that fetch those data and can be later used by all the data aggregator websites that has the $DSM token listed.

Reviewer: ```suggestion
Currently, inside our [CoinGecko](https://www.coingecko.com/en/coins/desmos) and [CoinMarketCap](https://coinmarketcap.com/currencies/desmos/) pages some important information about current and total supply are missing or not correctly updated. To solve this, we can implement a series of APIs that read those data from the chain and can be later used by all the data aggregator websites that have the token listed.
```

## Decision

> The APIs will be exposed in a new module called `x/supply` that will have the purpose to fetch the given coin info from different 
> cosmos-SDK modules (`x/bank`, `x/distribution`, `x/staking` ) and apply some conversions on them in order to avoid displaying them
> in millionth units.

Reviewer: ```suggestion
The APIs will be exposed in a new module called `x/supply` that will have the purpose to fetch the given information using different Cosmos SDK modules (namely `x/bank`, `x/distribution` and `x/staking` ) and apply some conversions on them in order to display them the best way possible for the client.
```

### Queries

#### Total Supply

> This query will:
> 1. Fetch the total supply of the given token `denom` from the `bank` module
> 2. Convert it in order to display its non-millionth value

Reviewer: ```suggestion
This query will fetch the total supply of a given token `denom`.
```

#### Current supply

> This query will:
> 1. Fetch the total supply of the given token `denom`
> 2. Fetch the total vested tokens amount
> 3. Fetch the community pool amount
> 4. Calculate the circulating supply by subtracting data of 2. and 3. from Total supply  

Reviewer: ```suggestion
This query will return the circulating supply amount by subtracting the total vested amount and the community pool amount from the total supply of the given token.
```

## Consequences

### Positive
- All the chain nodes will expose these APIs automatically after their node updates;
- No state-breaking changes introduced;
- Easily extensible.

## Test Cases

- Check API to correctly return the converted Total Supply from millionth to non-millionth representation;
- Check API to correctly return the current supply.

## References

> Actual alpha branch of `x/supply`:  
> https://github.com/desmos-labs/desmos/tree/leonardo/coingecko-APIs/x/supply.

Reviewer: It's better to put the issue containing the discussion here
```suggestion
- [Issue #733](https://github.com/desmos-labs/desmos/issues/773)
```
